### PR TITLE
My catalogue of things

### PR DESCRIPTION
**Pull request by mohamedgamaleldin999999**
In this PR, We are trying to build the ruby capstone project following these requirements:

[Logic]
Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
All methods visible in the diagram should be implemented:
add_item method (in all classes that have that method)
should take an instance of the Item class as an input.
should add the input item to the collection of items.
should add self as a property of the item object (by using the correct setter from the item object).
can_be_archived?() in the Item class
should return true if published_date is older than 10 years.
otherwise, it should return false.
move_to_archive() in the Item class
should reuse can_be_archived?() method.
should change the archived property to true if the result of the can_be_archived?() method is true.
should do nothing if the result of the can_be_archived?() method is false.
can_be_archived?() in the Book class
should override the method from the parent class.
should return true if parent's method returns true OR if cover_state equals to "bad".
otherwise, it should return false.
can_be_archived?() in the MusicAlbum class
should override the method from the parent class.
should return true if parent's method returns true AND if on_spotify equals true.
otherwise, it should return false.
can_be_archived?() in the Movie class
should override the method from the parent class.
should return true if parent's method returns true OR if silent equals true.
otherwise, it should return false.
can_be_archived?() in the Game class
should override the method from the parent class.
should return true if parent's method returns true AND if last_played_at is older than 2 years.
otherwise, it should return false.
Add unit tests for all implemented methods.
[User interface](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#user-interface)
Create a main.rb file that will serve as your console app entry-point.
Your console app, at the start, should:
Present the user with a list of options to perform.
Let users choose an option.
If needed, ask for parameters for the option.
Have a way to quit the app.
The following options should be available:
List all books
List all music albums
List all movies
List of games
List all genres (e.g 'Comedy', 'Thriller')
List all labels (e.g. 'Gift', 'New')
List all authors (e.g. 'Stephen King')
List all sources (e.g. 'From a friend', 'Online shop')
Add a book
Add a music album
Add a movie
Add a game
All data should be preserved by saving collections in .json files.
[Database](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#database)
Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.
[Project documentation](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#project-documentation)
Once you have finished the development of the project, you should record a video presenting the features of the project you built. It is a video with a maximum length of 5 minutes. The content of the video should include:

A description of the project.
A demo of the project features.
You should also highlight some interesting piece of code or something you built that you are very proud of.
You all should appear in the video and talk about the project.
For recording the video you can use the Zoom recording features while in a call with your peers.

[Workload distribution](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#workload-distribution)
In order to tackle this challenge, we created an almost empty [Kanban board template](https://github.com/microverseinc/curriculum-ruby/projects/1) - you can see only the task's placeholders. However, we grouped the above requirements into [lists for each member of your team](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things_divided.md). You will create your own copy of that template Kanban board and you will create tasks on your own in the separate activity.

You will be working in this way:

The group tasks (e.g., set up repo) will be divided among all of you or completed as a team (pair programming).
The other tasks related to the project will be developed individually. Every student will own tasks that will give a chance to practice a variety of skills.
All tasks should be based on the cards from your Kanban board.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,159 @@
...
+
+<!-- ACKNOWLEDGEMENTS -->
+
+## 🙏 Acknowledgments <a name="acknowledgements"></a>
+
+I would like to thank Microverse for including this in their curriculum.
+
+## 📝 License <a name="license"></a>
+
+This project is [MIT](./LICENSE) licensed.
```
Comment (cynthiainga) [inline README.md]: - [ ] Having broken links in the README file doesn’t look professional, kindly fix the broken link 😁 😁 

Screenshot 2023-10-20 at 23 15 47

File: Modules/book_module.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+require_relative '../classes/Book/book'
+
+module BookModule
+  def create_book(publisher, cover_state, publish_date)
+    book = Book.new(publisher, cover_state, publish_date)
+    @books << book
+    save_book_to_file(@books)
+  end
+
```
Comment (cynthiainga) [inline Modules/book_module.rb:9]: - [ ] When creating a book there is no way to add the label "title and color" which means the label can't be saved in the `label.json` if can't create it, kindly fix that 🙏🏾

File: Modules/game_module.rb
Comment (cynthiainga) [inline Modules/game_module.rb:1]: - [ ] When creating a game there is no way to add the `first and last name` of the author which means that data can't be saved in the `author.json` file if we can't create them, kindly fix that 🙏🏾 
- [ ] Also after creating a game the data isn’t preserved in the `game.json` file. The `multiplayer, last_played_at, publish_date, id` date isn’t preserved kindly fix that and make sure all data can be preserved 🙃

File: Modules/music_module.rb
Comment (cynthiainga) [inline Modules/music_module.rb:1]: - [ ] When creating music_album there is no way to add the genre of the music_album 'pop, jazz, hip pop,...' which means the genre can't be preserved in the `genre.json` file if we can't add it when creating music_album, kindly fix that as well 🙏🏾

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @mohamedgamaleldin999999, @obibaadoma & @waelelsafty07, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

- [ ] No linters errors ✔️ 
- [ ] Created `main.rb` file ✔️ 
- [ ] All tables are created ✔️ 
- [ ] Each class from the **UML class diagram** is created in a separate `.rb` file ✔️ 
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: JSON/game.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      "id": 780
+    },
+    {
+      "multiplayer": "yes",
+      "last_played_at": "2023-05-03",
+      "publish_date": "2023-05-03",
+      "id": 320
+    }
+  ]
+}
```
Comment (KayLemba) [inline JSON/game.json]: As suggested by the previous code reviewerWhen creating a book there is no way to add the label "title and color" which means the label can't be saved in the label.json if can't create it, kindly fix that

File: JSON/genre.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{
+    "Genres": [
+      {
+        "name": "pop"
+      }
+    ]
+  }
```
Comment (KayLemba) [inline JSON/genre.json]: As requested by the previous code reviewer when creating a game there is no way to add the first and last name of the author which means that data can't be saved in the author.json file if we can't create them, kindly fix that 🙏🏾
 Also after creating a game the data isn’t preserved in the game.json file. The multiplayer, last_played_at, publish_date, id date isn’t preserved kindly fix that and make sure all data can be preserved 🙃

File: JSON/label.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    {
+      "title": "Title 2",
+      "color": "green"
+    },
+    {
+      "title": "Title 3",
+      "color": "blue"
+    }
+  ]
+}
```
Comment (KayLemba) [inline JSON/label.json]: As suggested by the previous code reviewerWhen creating a book there is no way to add the label "title and color" which means the label can't be saved in the label.json if can't create it, kindly fix that it says in the project requirements that:
Preserving data

Verify if data is saved after quitting the app:

Open the console app and add a few items.
Close the app.
Open the app again and check it the items you added before are listed after using the correct option


Check if there is a method that saves data in .json files. That method should be used when user quits the app.
Check if data from the .json files is loaded when the app starts.

Make sure that app does not crashes if the .json file does not exist or is empty.

**Review by KayLemba — CHANGES_REQUESTED** [review]
Hi @mohamedgamaleldin999999, @obibaadoma & @waelelsafty07,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



Highlights

 No linters errors ✔️
 Created main.rb file ✔️
 All tables are created ✔️
 Each class from the UML class diagram is created in a separate .rb file ✔️
Required Changes ♻️

Check the comments under the review.

Optional suggestions

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code.

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

👍 You did well fixing the changes requested by the previous reviewers.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**